lib/node/pl-drivers/src/drivers/download_blob/sparse_cache/cache.ts

```diff
@@ -0,0 +1,153 @@
+  }
+
+  async get(key: string): Promise<Ranges> {
+    return await readRangesFile(this.logger, this.fPath(key));
+  }
+
+  async set(key: string, ranges: Ranges): Promise<void> {
+    return await writeRangesFile(this.fPath(key), ranges);
+  }
+
+  async delete(key: string): Promise<void> {
+    await fs.rm(this.fPath(key));
+  }
+}
+
+/** Extracted interface for storing sparse files. */
+export interface SparseFile {
+  exists(key: string): Promise<boolean>;
+  path(key: string): string;
+  write(key: string, data: Uint8Array, from: number): Promise<void>;
+  delete(key: string): Promise<void>;
+}
+
+/** Stores sparse files in a directory (the default implementation). */
+export class SparseCacheFsFile implements SparseFile {
+  constructor(
+    private readonly logger: MiLogger,
+    private readonly cacheDir: string,
+  ) {}
+
+  async exists(key: string): Promise<boolean> {
+    return await fileExists(this.path(key));
+  }
+
+  path(key: string): string {
+    return path.join(this.cacheDir, key);
+  }
+
+  async write(key: string, data: Uint8Array, from: number): Promise<void> {
+    await ensureDirExists(this.cacheDir);
+    await writeToSparseFile(this.logger, process.platform, this.path(key), data, from);
+  }
+
+  async delete(key: string): Promise<void> {
+    await fs.rm(this.path(key));
+  }
+}
+
+/** LRU cache for ranges of sparse files. */
+export class SparseCache {
+  /** Fields are public for tests. */
+
+  public keyToLastAccessTime = new Map<string, Date>();
+  public size = 0;
+
+  constructor(
+    public readonly logger: MiLogger,
+    public readonly maxSize: number,
+    public readonly ranges: SparseCacheRanges,
+    public readonly storage: SparseFile,
+  ) {}
+
+  public async initFromDir(cacheDir: string) {
+    await ensureDirExists(cacheDir);
+
+    const now = new Date();
+    for (const file of await fs.readdir(cacheDir)) {
+      if (file.endsWith(rangesFilePostfix)) {
+        continue;
+      }
+
+      const ranges = await this.ranges.get(file);
+      this.size += rangesSize(ranges);
+      this.keyToLastAccessTime.set(file, now);
+    }
+  }
+
+  async get(key: string, range: RangeBytes): Promise<string | undefined> {
+    if (await this.storage.exists(key)) {
+      this.keyToLastAccessTime.set(key, new Date());
+
+      const ranges = await this.ranges.get(key);
+      if (existRange(ranges, range)) {
+        return this.storage.path(key);
+      }
+
+      return undefined;
+    }
+
+    return undefined;
+  }
+
+  async set(key: string, range: RangeBytes, data: Uint8Array): Promise<void> {
+    await this.setNoClear(key, range, data);
+    await this.ensureCleared();
+  }
+
+  async setNoClear(key: string, range: RangeBytes, data: Uint8Array): Promise<void> {
+    this.keyToLastAccessTime.set(key, new Date());
+
+    const ranges = await this.ranges.get(key);
+    this.size -= rangesSize(ranges);
+
+    await this.storage.write(key, data, range.from);
```

Comment:
I'd set a safety check here to ensure you really write the data covered by range. When you write MORE than range it is safe. When you write less - it means next time you will think you wrote the data you really didn't.

Let's put sanity check here that data size is `==` to the range size (or at least `>=`).